rfcs/0079-enforce-prs.md

```diff
@@ -0,0 +1,80 @@
+---
+feature: No more direct pushes to master and release branches
+start-date: 2020-09-25
+author: Janne Heß <janne@hess.ooo>
+co-authors: Matthias Beyer <mail@beyermatthias.de>
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Enforce usage of Pull Requests for **all** contributions to nixpkgs master and release branches, which implies that these branches do not allow direct pushes anymore.
+
+This implements the four-eyes principle and allows easier change discussion, both before and after the merge, and improves overall security since each Pull Request has to be approved by at least one other person prior to merging.
+
+# Motivation
+[motivation]: #motivation
+
+In its current state every [Nixpkgs committer](https://github.com/orgs/NixOS/teams/nixpkgs-committers/members) is able to push arbitrary code to every branch, including master and all release branches.
+
+Pushing to critical branches branches of a software development project is not a good practice in more-than-one-contributor environments.
+Security vulnerabilities, breaking changes which are not tested, and regressions are easily introduced this way.
+Those issues can been prevented by a proper workflow and tooling that comes with the workflow.
+
+In the last year (as of 2020-09-25), we had
+- 19471 commits on the master branch in nixpkgs
+- of which 10347 (53.14%) were merges
+- of which 9124 (46.85%) were non-merge commits
+
+<small>
+(`git log --oneline --since="1 year ago" --first-parent [--[no-]merges] | wc -l`)
+</small>
+
+This is an improvement over the previous statistic from 2019-04-13 (where 51.85% commits went directly to master) and proves the point that committers are able to work properly without pushing directly.
+
+By changing our branching workflow to a no-push-to-master workflow, we can achieve more security, stability and even more important: better scalability.
+
+# Detailed design
+[design]: #detailed-design
+
+In GitHubs [branch protection](https://github.com/NixOS/nixpkgs/settings/branch_protection_rules) rules, branch protection rules which require pull request reviews, include administrators, forbid force pushes and branch deletions must be created.
+There must be rules for:
+- master
+- nixos-*
+- nixpkgs-*
+- release-*
+
```

Comment:
There should at least be some mention of what happens with other critical branches like "staging" and "staging-next". I don't know how these branches are updated currently, but I'd expect that workflows surrounding them will have to change.

Reply:
I agree. As far as I know, CI only ensures the `staging` branch builds before merging it, but not whether the actual merge builds, does it?
Because that would be (my) major concern here. Other than that, from what I know, I don't think the workflow should be changed. My concern with this PR is to get an evergreen master (and release branches), which (IMO) increases the development speed and scalability. Staging does not fall under that concern.